apis/filesystem/src/main/java/org/jclouds/filesystem/strategy/internal/FilesystemStorageStrategyImpl.java

```diff
@@ -254,15 +255,49 @@ public void clearContainer(final String container) {
    public void clearContainer(String container, ListContainerOptions options) {
       filesystemContainerNameValidator.validate(container);
       // TODO: these require calling removeDirectoriesTreeOfBlobKey
-      checkArgument(options.getDir() == null && options.getPrefix() == null, "cannot specify directory or prefix");
+//      checkArgument(options.getDir() == null && options.getPrefix() == null, "cannot specify directory or prefix");
+      String optsPrefix;
+      // Pick whichever one is not null? Not sure what to do until inDirectory is deprecated.
+      optsPrefix = options.getDir() == null ? options.getPrefix() : options.getDir();
+      // If both are null, just use an empty string
+      optsPrefix = optsPrefix == null ? "" : optsPrefix;
+      String normalizedOptsPath = normalize(optsPrefix);
+      String basePath = buildPathStartingFromBaseDir(container, normalizedOptsPath);
+      filesystemBlobKeyValidator.validate(basePath);
       try {
-         File containerFile = openFolder(container);
-         File[] children = containerFile.listFiles();
-         if (null != children) {
-            for (File child : children)
-               if (options.isRecursive() || child.isFile()) {
-                  Utils.deleteRecursively(child);
+         File object = new File(basePath);
+         if (object.isFile()) {
+            // To mimic the S3 type blobstores, a prefix for an object blob
+            // should also get deleted
+            delete(object);
+         }
+         else if (object.isDirectory() & (optsPrefix.endsWith(File.separator) | isNullOrEmpty(optsPrefix))) {
+            // S3 blobstores will only match prefixes that end with a trailing slash/file separator
+            // For insance, if we have a blob at /path/1/2/a, a prefix of /path/1/2 will not list /path/1/2/a
+            // but a prefix of /path/1/2/ will
+            File containerFile = openFolder(container + File.separator + normalizedOptsPath);
+            File[] children = containerFile.listFiles();
+            if (null != children) {
+               for (File child : children) {
+                  if (options.isRecursive()) {
+                     Utils.deleteRecursively(child);
+                  } else {
+                     if (child.isFile()) {
+                        Utils.delete(child);
```

Comment:
It seems like both this and `removeDirectoriesTreeOfBlobKey` have similar logic -- can you consolidate these?

Reply:
It might be possible, but `removeDirectoriesTreeOfBlobKey` walks up the tree, and this one is walking deeper into the tree. Suggestions?

Reply:
I guess this is the best we can do.